core/src/com/google/inject/internal/aop/ImmutableStringTrie.java

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (C) 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.inject.internal.aop;
+
+import static java.util.Arrays.binarySearch;
+
+import java.util.Collection;
+import java.util.function.ToIntFunction;
+
+/**
+ * Immutable space-efficient trie that provides a quick lookup index for a sorted array of strings.
```

Comment:
nit, "a sorted set of non empty strings" since this expect the array to not have duplicate strings and will always return "-1" for empty strings.

Reply:
92e81c7